Use Spyder's new PythonpathManager plugin

jitseniesen opened the pull request:
This PR allows the plugin to run with Spyder 6, so the PR also bumps the plugin version and the dependency to specify that Spyder 6 is installed with the plugin.

Fixes #195 
Fixes #167

jitseniesen committed:
Use Spyder's new PythonpathManager plugin

This is necessary to support Spyder 6

jitseniesen committed:
Depend on Spyder v6 and bump version to 0.6.0.dev0

jitseniesen committed:
Fix Spyder installation from Git in CI test

Specifically, if Spyder is installed from Git, then don't try to
also install it with conda.

jitseniesen committed:
Don't install Spyder via conda in CI test

The plugin currently depends on Spyder 6 which is not yet released.

ccordoba12 commented on spyder_unittest/unittestplugin.py:
```diff
@@ -148,6 +148,14 @@ def on_projects_available(self):
+        """
+        Connect to signal announcing that Python path changed.
+        """
+        pythonpathManager = self.get_plugin(Plugins.PythonpathManager)
+        pythonpathManager.sig_pythonpath_changed.connect(self.update_pythonpath)
```
```suggestion
        ppm = self.get_plugin(Plugins.PythonpathManager)
        ppm.sig_pythonpath_changed.connect(self.update_pythonpath)
```

This is the acronym we decided to use for the Pythonpath manager in Spyder.

ccordoba12 reviewed (commented):
Thanks @jitseniesen for your work on this!

jitseniesen committed:
Fix code style

Co-authored-by: Carlos Cordoba <ccordoba12@gmail.com>

jitseniesen commented on spyder_unittest/tests/test_unittestplugin.py:
```diff
@@ -38,6 +38,7 @@ def test_plugin_initialization(plugin):
     assert plugin.main.run_menu_actions[1].text() == 'Run unit tests'
 
 
+@pytest.mark.skip('not clear how to test interactions between plugins')
```
> What do you want to check with this test? I can give you a hand to fix it if I understand better what it does.

Broadly speaking, I want to check that if the Python path changes in the Python Path Manager, this change is picked up by the unittest plugin. Can you give me an example of a test that tests the interaction between two plugins? I should be able to take it from there.

jitseniesen replied:
> What do you want to check with this test? I can give you a hand to fix it if I understand better what it does.

Broadly speaking, I want to check that if the Python path changes in the Python Path Manager, this change is picked up by the unittest plugin. Can you give me an example of a test that tests the interaction between two plugins? I should be able to take it from there.

ccordoba12 replied:
> Can you give me an example of a test that tests the interaction between two plugins?

When we need to do that, we add the test to `spyder/app/tests/test_mainwindow.py`. So, I'd suggest to add here a simplified version of the `main_window` fixture we have in `spyder/app/tests/conftest.py`.

That shouldn't be that hard since you don't need to test as many things as in Spyder. But I think it's important because, as far as I can see, this plugin's connections to others are basically untested (i.e. all tests in this file are skipped now).

jitseniesen committed:
Add plugin teardown methods

jitseniesen committed:
Add integration tests

ccordoba12 commented on spyder_unittest/tests/conftest.py:
```diff
@@ -0,0 +1,55 @@
+os.environ['SPYDER_PYTEST'] = 'True'
+
+# Spyder imports
+from spyder.app.tests.conftest import main_window
```
We don't package our tests as part of the final package to save space. That's why I mentioned in my previous comment that you need to create your own `main_window` fixture here.

jitseniesen replied:
Thanks, I missed that. You also suggested that I simplify the fixture, which sounds a bit scary because there are some places where it is not immediately clear to me what is going on, but I will have a go at it.

jitseniesen committed:
Use simplified main_window fixture

Do not import the fixture from Spyder because tests (and thus fixtures)
are not distributed with Spyder.

jitseniesen committed:
Disable loading of old third-party plugins in Spyder

This seems to cause a segmentation fault when pytest exits,
for reasons that are not clear. Perhaps the code should be
removed from Spyder.

jitseniesen committed:
Remove monkeypatching from previous commit

This is no longer necessary now that Spyder PR #20789 is merged.

jitseniesen commented:
@ccordoba12 Thanks for resolving the issue in Spyder so quickly. I changed the test fixture to reflect your changes and I think it is ready now. Let me know whether you want to do another review or want me to merge.

ccordoba12 commented on spyder_unittest/tests/test_unittestplugin.py:
```diff
@@ -1,128 +1,165 @@
-        plugin.handle_project_change)
-    plugin.main.projects.sig_project_closed.connect.assert_called_with(
-        plugin.handle_project_change)
+    projects._create_project(project_dir)
```
Should we make `_create_project` a public method in the Projects plugin API? It looks like the appropriate method to call to easily create projects.

jitseniesen replied:
@ccordoba12 Yes, it makes sense to make `_create_project` public. This does commit us to some API stability which may hinder development of projects in the near future; you should have a decent guess whether this is likely to be a problem. The name is also quite similar to the `create_new_project` method but I don't have a solution for that.

ccordoba12 replied:
Ok, thanks for the feedback.

ccordoba12 reviewed (approved):
I left a simple question, the rest looks good to me @jitseniesen.

And great work with the integration tests! They are really cool.
